### PR TITLE
SCA Updates - PaymentIntents & SetupIntents management

### PR DESCRIPTION
**Pull request by jeremypele**
Android SDK 10.2.1
iOS SDK 16.0.6

Removal of `sources` in favor of `payment_method` and `payment_intent`
3DS2 verifications
Automatic and manual validation flows

### CODE REVIEW

**Comment by jeremypele** [comment]
@RobertGardner PR Completed :)

File: README.md [hunk 1]
```diff
@@ -84,36 +84,6 @@ const cc = new Card("1111111111111111",2,18,"123");
 cc.name = "Osei Fortune";
 ```
 
-### Get Token
-
```
Comment (jeremypele) [inline README.md:88]: Removed example as token creation is now deprecated (even if it's still supported)

File: demo-angular/app/demo/intent.component.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    this.stripe = new Stripe(publishableKey);
+  }
+
+  registerCard(cardView: CreditCardView) {
+    this._setStatus("Create Setup Intent...");
+    this.stripeService.createSetupIntent().then((intent) => {
+
+      this._setStatus("Create Payment Method...");
+      this.stripe.createPaymentMethod(cardView.card, (error, pm) => {
```
Comment (RobertGardner) [inline demo-angular/app/demo/intent.component.ts:50]: Do you need to handle the error here?

File: src/standard/index.d.ts [hunk 3]
```diff
@@ -37,7 +37,17 @@ export declare interface StripeBackendAPI {
    * @returns a Promise that resolves on success and rejects on failure.
    *     Any error should be reported as a string that can be displayed to the user.
    */
-  completeCharge(sourceID: string, amount: number, shippingMethod?: StripeShippingMethod, shippingAddress?: StripeAddress): Promise<void>;
+  capturePayment(sourceID: string, amount: number, shippingMethod?: StripeShippingMethod, shippingAddress?: StripeAddress): Promise<void>;
+
+
```
Comment (RobertGardner) [inline src/standard/index.d.ts]: Should we keep `completeCharge` for backward compatibility?

(also, extra blank line)

File: src/standard/index.d.ts [hunk 4]
```diff
@@ -37,7 +37,17 @@ export declare interface StripeBackendAPI {
...
+  /**
+   * Calls the client-implemented Stripe backend to confirm a Payment Intent (manual confirmation mode).
+   *
+   * @param paymentIntentID The Payment Intent ID to send to the backend.
+   * @returns a Promise that resolves on success and rejects on failure.
+   *     Any error should be reported as a string that can be displayed to the user.
+   */
+  confirmPaymentIntent(paymentIntentID: string): Promise<any>;
```
Comment (RobertGardner) [inline src/standard/index.d.ts]: Do we know what type it returns in the promise? I'd prefer not using `any` if it can be helped.

File: src/standard/standard.android.ts [hunk 5]
```diff
@@ -50,13 +56,19 @@ export class StripeConfig extends StripeConfigCommon {
...
   native: com.stripe.android.CustomerSession;
 
   constructor() {
-    com.stripe.android.CustomerSession.initCustomerSession(createKeyProvider());
+    StripeConfig.shared().initPaymentConfiguration();
+    com.stripe.android.CustomerSession.initCustomerSession(this._getContext(), createKeyProvider());
     this.native = com.stripe.android.CustomerSession.getInstance();
   }
+
+  private _getContext(): android.content.Context {
```
Comment (RobertGardner) [inline src/standard/standard.android.ts]: Minor stylistic point. Personally, I prefer
`private get context(): android.context.Context {}`

which I think looks better at the call site: `this.context`.

File: src/standard/standard.android.ts [hunk 6]
```diff
@@ -113,7 +126,9 @@ export class StripePaymentSession {
         .build();
     }
     this.native = new com.stripe.android.PaymentSession(this.patchActivity());
-    if (!this.native.init(createPaymentListener(this, listener), config)) {
+
```
Comment (RobertGardner) [inline src/standard/standard.android.ts:129]: Accidental extra blank lines?

File: src/stripe.ios.ts [hunk 7]
```diff
@@ -1,4 +1,6 @@
 import { View } from 'tns-core-modules/ui/core/view';
+import { topmost } from "tns-core-modules/ui/frame";
```
Comment (RobertGardner) [inline src/stripe.ios.ts:2]: I believe this is now in master. You may need to rebase.

**Review by RobertGardner — COMMENTED** [review]
Wow, that was a large effort! Thank you so much for doing it!

I made a few minor comments and am now running it to test it out.

**Review by RobertGardner — COMMENTED** [review]
On iOS the Standard Demo and Custom Demo worked (Angular). I was actually not expecting it to. :-)

The Intent demo worked also, but it wasn't clear to me exactly what steps I was supposed to take. I also expected to see more debug information -- did it get clipped off? When I clicked Manual Confirm the screen went black and I got an error:

```
*** Assertion failure in -[STPAPIClient(PaymentIntents) retrievePaymentIntentWithClientSecret:completion:](), /Users/rgardner/Development/HearingClinic/triniwiz/nativescript-stripe/demo-angular/platforms/ios/Pods/Stripe/Stripe/STPAPIClient.m:706
```

I'm not sure what was supposed to happen, though. Maybe we need to document this somewhere.

We might need to document all the Intents stuff and what to do to make manual confirmation work.

**Review by RobertGardner — COMMENTED** [review]
While testing I came across this error when clicking Setup Intent:
```
JS: ERROR Error: Uncaught (in promise): Error: <h1>Not Found</h1>
JS: Error: <h1>Not Found</h1>
JS:     at file:///app/demo/stripe.service.ts:97:14
```
From this and other problems I'm seeing, I'm gathering that the backend this expects to call has also changed with this version. 

Oh, just looked at it in github and, indeed, it's new. I'll deploy it and try some more. I wonder though if we should put something in the CHANGELOG that says the example backend has changed and needs to be updated.

File: src/standard/index.d.ts (hunk 3 above)
Comment (jeremypele) [reply src/standard/index.d.ts]: Yes good point

Comment (jeremypele) [reply src/standard/index.d.ts]: I'd rather mention it in the Changelog, but I don't think it's a breaking change for anyone as it's wrapped into specific methods on every calls. 
Only case would be if you hit `stripeService. completeCharge` directly (and I don't think it's happening out of a sandbox mode)

**Comment by jeremypele** [comment]
> Oh, just looked at it in github and, indeed, it's new. I'll deploy it and try some more. I wonder though if we should put something in the CHANGELOG that says the example backend has changed and needs to be updated.

Ah ok, didn't work with the previous version of the demo back-end app. Will put a note on the Changelog

 > On iOS the Standard Demo and Custom Demo worked (Angular). I was actually not expecting it to. :-)

Yes, tried to make sure that previous flows will continue to work (even though It won't always complete the payement if you're using a 3DS2 credit card)

 > also expected to see more debug information -- did it get clipped off?

What kind of debug information were you expecting/looking to find?

File: src/stripe.ios.ts (hunk 7 above)
Comment (jeremypele) [reply src/stripe.ios.ts:2]: Just check and it's not actually

**Comment by jeremypele** [comment]
> I'm not sure what was supposed to happen, though. Maybe we need to document this somewhere.
> 
> We might need to document all the Intents stuff and what to do to make manual confirmation work.

Yes, took me some time to figure the whole different flows out.

Not sure that we should explain the different intents logic in the plugin description as we're only mapping the in-place SDKs. That's the reason I've put the links to [manual](https://stripe.com/docs/payments/payment-intents/ios#handle-authentication-manual) and [automatic](https://stripe.com/docs/payments/payment-intents/ios#automatic-confirmation-ios) flows above each method in the component.

Basically, `SetupIntent` is the way to register a new Card in order to charge it later.
As for `PaymentIntent`, the UI holds all the paiement logic on the `automatic` flow whereas the UI only authenticate the intent and let the back-end actually charge the credit card (or any other payment method you used).

**Comment by jeremypele** [comment]
> ```
> *** Assertion failure in -[STPAPIClient(PaymentIntents) retrievePaymentIntentWithClientSecret:completion:](), /Users/rgardner/Development/HearingClinic/triniwiz/nativescript-stripe/demo-angular/platforms/ios/Pods/Stripe/Stripe/STPAPIClient.m:706
> ```

Could not replicate this issue on my end, did you manage to found what's the issue is?

**Comment by carlosanjos** [comment]
I have a `CreditCardView` inside of a modal window `ModalDialogService`. The following error happens when I try to confirm a setup intent `stripe.confirmSetupIntent()`  ***Error Domain=STPPaymentHandlerErrorDomain Code=7 "There was an unexpected error -- try again in a few seconds" UserInfo={com.stripe.lib:ErrorMessageKey=The authenticationContext is invalid.  Make sure it's non-nil and in the window hierarchy., NSLocalizedDescription=There was an unexpected error -- try again in a few seconds}*** I can share some snippets if I am not clear enough.

**Comment by jeremypele** [comment]
> I have a `CreditCardView` inside of a modal window `ModalDialogService`. The following error happens when I try to confirm a setup intent `stripe.confirmSetupIntent()` _**Error Domain=STPPaymentHandlerErrorDomain Code=7 "There was an unexpected error -- try again in a few seconds" UserInfo={com.stripe.lib:ErrorMessageKey=The authenticationContext is invalid. Make sure it's non-nil and in the window hierarchy., NSLocalizedDescription=There was an unexpected error -- try again in a few seconds}**_ I can share some snippets if I am not clear enough.

@carlosanjos Won't have time to review this this week but as the hostViewController is the nativescript root page, it's possible that modals are not supported as the displayed VC layer is not the one that hosts the context

in `stripe.ios.ts`
```
private _getAuthentificationContext(): STPPaymentContext {
    const authContext = STPPaymentContext.alloc();
    authContext.hostViewController = topmost().currentPage.ios; // here lies the issue
```

File: src/standard/index.d.ts (hunk 4 above)
Comment (RobertGardner) [reply src/standard/index.d.ts]: So I was trying to figure this out and discovered that `confirmPaymentIntent` is never called by the plugin itself. That makes me wonder why this call was added to `StripeBackendAPI` instead of just being a method implemented in the demo apps' `StripeService`.

One argument is that it is a Stripe backend call. The other argument is that `StripeBackendAPI` represents the contract between the plugin and the client -- it has all the calls that the plugin requires the client to implement.

Thoughts?

Comment (jeremypele) [reply src/standard/index.d.ts]: Yes, that's a `StripeService` backend call - needs to be declared on the interface to be callable from the user flow.

**Comment by carlosanjos** [comment]
> > I have a `CreditCardView` inside of a modal window `ModalDialogService`. The following error happens when I try to confirm a setup intent `stripe.confirmSetupIntent()` _**Error Domain=STPPaymentHandlerErrorDomain Code=7 "There was an unexpected error -- try again in a few seconds" UserInfo={com.stripe.lib:ErrorMessageKey=The authenticationContext is invalid. Make sure it's non-nil and in the window hierarchy., NSLocalizedDescription=There was an unexpected error -- try again in a few seconds}**_ I can share some snippets if I am not clear enough.
> 
> @carlosanjos Won't have time to review this this week but as the hostViewController is the nativescript root page, it's possible that modals are not supported as the displayed VC layer is not the one that hosts the context
> 
> in `stripe.ios.ts`
> 
> ```
> private _getAuthentificationContext(): STPPaymentContext {
>     const authContext = STPPaymentContext.alloc();
>     authContext.hostViewController = topmost().currentPage.ios; // here lies the issue
> ```

@jeremypele I also tried having getting the context from the root controller, but I've got the same problem :(

```
class StripeAuthenticationContext extends NSObject implements STPAuthenticationContext {
  public static ObjCProtocols = [ STPAuthenticationContext ];

  authenticationPresentingViewController(): UIViewController {
    return UIApplication.sharedApplication.keyWindow.rootViewController;
  }
}```

**Comment by jeremypele** [comment]
@carlosanjos I've fixed it on my last commit and added a modal example on the Intent angular demo to test it

File: src/standard/index.d.ts (hunk 4 above)
Comment (RobertGardner) [reply src/standard/index.d.ts]: That explains why the call is in `StripeService` but not why it's in `StripeBackendAPI`.

I removed it from the `StripeBackendAPI` interface, made no other changes, and rebuilt and ran. It builds and everything seems to work correctly.

**Comment by RobertGardner** [comment]
You've done outstanding work on this. Thanks again!

I updated the backend on Heroku and the iOS version now works superbly. All the buttons work, walk through the stages, show success messages. It's great.

I'm having trouble on Android, though. When I click any of the Intents buttons the app gives a message like "Setting up Intent..." and then freezes. Eventually I'll get an error in the log (and a blank alert) that says:
```
JS: ERROR Error: Uncaught (in promise): Error: java.net.UnknownHostException: Unable to resolve host "ns-stripe-example-backend.herokuapp.com": No address associated with hostname
```

The same settings work on iOS, so there's something strange happening on Android.

(I think the reason the alert is blank is because I also see this: 
```
JS: ERROR Error: Uncaught (in promise): Error: Cannot convert object to Ljava/lang/String; at index 1
JS: Error: Cannot convert object to Ljava/lang/String; at index 1
JS:     at file:///app/Users/rgardner/Development/HearingClinic/triniwiz/nativescript-stripe/src/standard/standard.android.ts:81:28
```

**Comment by jeremypele** [comment]
Thanks for the kind words @RobertGardner!

Regarding the issue on Android, I'm not facing them. Silly question but did you put the ` https://` prefix in your `backendBaseURL`? And do your simulator has access to internet?
Sounds silly but I've had those kind of issue in production in some android apps when internet connection is not available.

The issue you've pinned refered to this line, there's an issue with the customer session key
```function createKeyProvider(): com.stripe.android.EphemeralKeyProvider {
  return new com.stripe.android.EphemeralKeyProvider({
    createEphemeralKey(apiVersion: string, keyUpdateListener: com.stripe.android.EphemeralKeyUpdateListener): void {
      StripeConfig.shared().backendAPI.createCustomerKey(apiVersion)
        .then(key => {
          keyUpdateListener.onKeyUpdate(JSON.stringify(key));
        }).catch(e => {
        >>>>  keyUpdateListener.onKeyUpdateFailure(500, e);
        });
    }
  });
}```

**Comment by RobertGardner** [comment]
Well, I guess I'm not sure why it didn't work before. Most likely I didn't have internet access like I thought I did because when I tried it again, it worked like a charm!

The error on the call to `keyUpdateListener.onKeyUpdateFailure(500, e);` looks like an error that was in there before, so not your fault. The problem is that `onKeyUpdateFailure` expects a string as the second parameter, but if `StripeService._postRequest` gets an exception from `httpModule.request` then it throws a non-string value, which then gets sent to `onKeyUpdateFailure`'s second parameter. Unfortunately, I was new to TypeScript/JavaScript when I wrote the `StripeBackendAPI` spec and didn't realize that throwing a string was non-ideal.

So, the question is, do we fix it now so they are spec'd to actually throw Error (and fix the call sites) or do we just let it go? I'm inclined to fix it (I can do that if you'd prefer) because we're bumping the major version so we can introduce incompatibilities. On the other hand, this isn't a huge deal and fixing `createKeyProvider` would be simple (simply change the second parameter in the call to `onKeyUpdateFailure` to `JSON.stringify(e)`).

What is your opinion?

**Comment by RobertGardner** [comment]
Another comment: This is getting real close. We just need to clean up the outstanding comments and then, I believe, it will be ready to submit. Do you agree?

**Comment by jeremypele** [comment]
Yes, I'd fix the issue. It should not block this release as it's not related, but as we have the solution it'd be stupid not to do that now.
I'll push the fix soon, and yes it should be ready to be promoted

**Comment by RobertGardner** [comment]
Hey, I keep looking at this and thinking it must be close. Could you let me know when you're ready for my final review and merge? Thanks!

**Comment by jeremypele** [comment]
@RobertGardner Yes, ready for a last review. I'm already using a tarball from the local fork on my app FYI

File: src/standard/standard.common.ts [hunk 8]
```diff
@@ -53,7 +53,16 @@ export interface StripeBackendAPI {
...
+  capturePayment(sourceID: string, amount: number, shippingMethod?: StripeShippingMethod, shippingAddress?: StripeAddress): Promise<void>;
+
+  /**
+   * Calls the client-implemented Stripe backend to confirm a Payment Intent (manual confirmation mode).
+   *
+   * @param paymentIntentID The Payment Intent ID to send to the backend.
+   * @returns a Promise that resolves on success and rejects on failure.
+   *     Any error should be reported as a string that can be displayed to the user.
+   */
+  confirmPaymentIntent(paymentIntentID: string): Promise<any>;
```
Comment (RobertGardner) [inline src/standard/standard.common.ts]: Remove the method here as well.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,19 @@
...
+### Updates
+- SDK Updates for SCA compliance
+
+### Implements
+- [(# 78)](https://github.com/triniwiz/nativescript-stripe/issues/78) Stripe SDKs out of date; Upgrade needed for latest SCA process
+- [(# 45)](https://github.com/triniwiz/nativescript-stripe/issues/45) EU Strong Customer Authentication support integration
+
+### Breaking Changes
+- Sources and Tokens are now deprecated, uses Payment Methods and Payment Intents instead
+- [Back-end test app](https://github.com/stripe/example-ios-backend) needs to be updated if you want to plug with Intents.
```
Comment (RobertGardner) [inline CHANGELOG.md]: "if you want to plug with Intents" -> "if you want to use the plugin with Intents"   ???

File: README.md [hunk 10]
```diff
@@ -149,29 +119,23 @@ for some credit card purchases. Stripe supports this, though most of the work to
...
 * `StripePaymentIntent` and related classes
-* Methods `Stripe.createPaymentMethod`, `Stripe.retrievePaymentIntent`, and `Stripe.confirmPaymentIntent`
+* Methods `Stripe.createPaymentMethod`, `Stripe.retrievePaymentIntent`, `Stripe.confirmPaymentIntent` and `Stripe.confirmSetupIntent`
 
 ## Handling secondary customer input
 SCA requires the customer to enter additional information with some charge cards. Stripe takes care of this
 if you properly handle the redirect from the `StripePaymentIntent` returned from the server.
 
-On iOS, `StripeRedirectSession` can help manage the interaction with the customer.
+If you're using the [automatic confirmation flow](https://stripe.com/docs/payments/payment-intents/ios#automatic-confirmation-ios), `confirmPaymentIntent` and `confirmSetupIntent` will automatically manages the SCA validation by showing and validating the payment authentification.
```
Comment (RobertGardner) [inline README.md]: "manages" -> "manage"

File: README.md [hunk 11]
```diff
@@ -149,29 +119,23 @@ for some credit card purchases. Stripe supports this, though most of the work to
...
-cannot be called on the UI thread in Android (but they can on iOS), and the technique for handling
-secondary customer interaction is difficult. Hopefully these will be addressed soon.
+If you're using the [manual confirmation flow](https://stripe.com/docs/payments/payment-intents/ios#manual-confirmation-ios), where back-end creates the `PaymentIntent`|`SetupIntent` and requires an Intent authentification from the customer, `authenticatePaymentIntent` and `authenticateSetupIntent` will allow to manage that extra step before sending back the Intent to your server.
 
 ## Status
-Finally, keep in mind that while these SCA support classes and methods have been added to the plugin, they
-have not yet been thoroughly tested. The `demo` and `demo-angular` apps do *not* (yet) support `PaymentIntent`,
-although they have been modified to work with the current Stripe example backend
-(which does fully support `PaymentIntent`). Any help updating the demo apps would be greatly appreciated.
+`demo-angular` now support `SetupIntent` and `PaymentIntent` SCA integration. Any credit card verification will be automatically prompted to the user.
```
Comment (RobertGardner) [inline README.md]: "support" -> "supports"

File: demo/app/demo/stripe.service.ts [hunk 12]
```diff
@@ -36,48 +36,78 @@ export class StripeService implements StripeBackendAPI {
...
+    paymentSession.requestPayment();
   }
 
-  completeCharge(stripeID: string, amount: number, shippingMethod: StripeShippingMethod, shippingAddress: StripeAddress): Promise<void> {
-    let url = this.backendURL("capture_payment");
+  /*
+   *  Private
+   */
+
+  private _postRequest(endpoint: string, content: string = ''): Promise<any> {
```
Comment (RobertGardner) [inline demo/app/demo/stripe.service.ts]: Trivial: For consistency, should probably use " instead of '.

When I started this plugin, Angular was recommending double quotes. Looks like they've now switched to single quotes. I should, too, I guess... sometime.

**Review by RobertGardner — COMMENTED** [review]
Awesome! Some trivial changes and it'll be ready to merge.

File: demo/app/demo/stripe.service.ts (hunk 12 above)
Comment (jeremypele) [reply demo/app/demo/stripe.service.ts]: Yup, Only using `'` on my projects, try to stick to the author conventions when doing some PR but old habits die hard 😂

**Comment by jeremypele** [comment]
@RobertGardner I've pushed the last updates (was away for the weekend)
